File: flutter_app/test/services/prescription_handover_test.dart
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('Prescription Handover Flow', () {
    test('should create valid confirmation payload', () {
      final payload = {'driver_id': 1};

      expect(payload['driver_id'], 1);
    });

    test('should validate driver_id is required', () {
      final payload = <String, dynamic>{};

      final isValid = payload['driver_id'] != null;
      expect(isValid, false);
    });

    test('should handle prescription confirmed state', () {
      final orderState = {'id': 1, 'prescription_confirmed': true};

      expect(orderState['prescription_confirmed'], true);
    });

    test('should handle prescription not confirmed state', () {
      final orderState = {'id': 1, 'prescription_confirmed': false};

      expect(orderState['prescription_confirmed'], false);
    });

    test('should block pharmacy address when not confirmed', () {
      final canViewAddress = false;
      final shouldBlock = !canViewAddress;

      expect(shouldBlock, true);
    });

    test('should allow pharmacy address when confirmed', () {
      final canViewAddress = true;
      final shouldBlock = !canViewAddress;

      expect(shouldBlock, false);
    });
  });

  group('Handover Checkbox Logic', () {
    test('should default to unchecked', () {
      const isChecked = false;
      expect(isChecked, false);
    });

    test('should toggle to checked', () {
      var isChecked = false;
      isChecked = true;
      expect(isChecked, true);
    });

    test('should require confirmation before proceeding', () {
      const userConfirmed = false;
      const canProceed = userConfirmed;

      expect(canProceed, false);
    });

    test('should allow proceeding after confirmation', () {
      const userConfirmed = true;
      const canProceed = userConfirmed;

      expect(canProceed, true);
    });
  });

  group('Edge Cases', () {
    test('should handle null driver_id', () {
      final payload = {'driver_id': null};

      expect(payload['driver_id'], isNull);
    });

    test('should handle invalid driver_id', () {
      final payload = {'driver_id': -1};

      expect(payload['driver_id'], lessThan(0));
    });

    test('should handle double confirmation attempt', () {
      var confirmed = false;
      confirmed = true;

      final canConfirmAgain = !confirmed;
      expect(canConfirmAgain, false);
    });

    test('should handle invalid order ID', () {
      const orderId = 'invalid';
      final parsed = int.tryParse(orderId);

      expect(parsed, isNull);
    });

    test('should handle negative order ID', () {
      const orderId = -1;
      final isValid = orderId > 0;

      expect(isValid, false);
    });

    test('should handle zero order ID', () {
      const orderId = 0;
      final isValid = orderId > 0;

      expect(isValid, false);
    });
  });

  group('UI State Management', () {
    test('should show loading while confirming', () {
      const isLoading = true;
      expect(isLoading, true);
    });

    test('should hide loading after confirmation', () {
      const isLoading = false;
      expect(isLoading, false);
    });

    test('should show error on failure', () {
      final errorState = {
        'hasError': true,
        'message': 'Failed to confirm prescription',
      };

      expect(errorState['hasError'], true);
    });

    test('should clear error on retry', () {
      var errorState = <String, dynamic>{'hasError': true, 'message': 'Error'};
      errorState = {'hasError': false, 'message': ''};

      expect(errorState['hasError'], false);
    });
  });
}

File: backend/tests/prescription_handover.test.js
const request = require('supertest');
const app = require('../src/index.js');

describe('Prescription Handover (Phase 32)', () => {
  describe('POST /api/orders/:id/confirm-prescription', () => {
    test('should return 404 for non-existent order', async () => {
      const res = await request(app)
        .post('/api/orders/99999/confirm-prescription')
        .send({ driver_id: 1 });

      expect(res.status).toBe(404);
    });

    test('should require driver_id', async () => {
      const res = await request(app)
        .post('/api/orders/1/confirm-prescription')
        .send({});

      expect(res.status).toBe(400);
    });

    test('should accept valid confirmation request', async () => {
      const res = await request(app)
        .post('/api/orders/1/confirm-prescription')
        .send({ driver_id: 1 });

      expect(res.status).toBe(200);
      expect(res.body).toHaveProperty('success');
    });
  });

  describe('GET /api/orders/:id/pharmacy-address (with handover check)', () => {
    test('should return 400 for invalid order ID', async () => {
      const res = await request(app)
        .get('/api/orders/invalid/pharmacy-address');

      expect(res.status).toBe(400);
    });

    test('should reject non-numeric order ID', async () => {
      const res = await request(app)
        .get('/api/orders/-1/pharmacy-address');

      expect(res.status).toBe(400);
    });
  });

  describe('Edge Cases', () => {
    test('should handle invalid order ID on confirm', async () => {
      const res = await request(app)
        .post('/api/orders/invalid/confirm-prescription')
        .send({ driver_id: 1 });

      expect(res.status).toBe(400);
    });

    test('should reject null driver_id', async () => {
      const res = await request(app)
        .post('/api/orders/1/confirm-prescription')
        .send({ driver_id: null });

      expect(res.status).toBe(400);
    });

    test('should reject missing driver_id', async () => {
      const res = await request(app)
        .post('/api/orders/1/confirm-prescription')
        .send({});

      expect(res.status).toBe(400);
    });
  });
});